.github/workflows/docs-preview-deploy.yml

```diff
@@ -17,12 +17,12 @@ jobs:
     name: 'Deploy Preview'
     runs-on: ubuntu-22.04
     # Requires a PR event triggered `docs-preview-prepare.yml` workflow run that was successful + ensure the head SHA belongs to an associated PR:
+    # NOTE: A multi-line `if` GHA expression must avoid wrapping with `${{ }}`, otherwise it is unintentionally parsed as a string:
+    # https://github.com/nikitastupin/pwnhub/blob/main/writings/if-condition.md
     if: |
-      ${{
-        github.event.workflow_run.conclusion == 'success'
-        && github.event.workflow_run.event == 'pull_request'
-        && contains(github.event.workflow_run.pull_requests.*.head.sha, github.event.workflow_run.head_sha)
-      }}
+      github.event.workflow_run.conclusion == 'success'
+      && github.event.workflow_run.event == 'pull_request'
+      && contains(github.event.workflow_run.pull_requests.*.head.sha, github.event.workflow_run.head_sha)
```

Comment:
This condition appears to be [triggering a skip](https://github.com/docker-mailserver/docker-mailserver/actions/runs/11760837244/job/32761886447) for some reason.

---

`docs-preview-prepare` ran successfully on the [PR pre-merge commit](https://github.com/docker-mailserver/docker-mailserver/actions/runs/11760832296/job/32761878540#step:2:59):

```console
$ git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 origin +a721146229a6adfb6f18cd28f14b8396c54ec6f7:refs/remotes/pull/4183/merge

a721146229a6adfb6f18cd28f14b8396c54ec6f7 -> pull/4183/merge
```

When cloning this repo and checkout that PR pre-merge commit, we can run `git log`:

```console
$ git log

commit a721146229a6adfb6f18cd28f14b8396c54ec6f7 (grafted, HEAD, pull/4183/merge)
Author: RoelSG <email-address-here>
Date:   Sun Nov 10 00:56:27 2024 +0000

    Merge a5682e7c805397493aec6e5f333b236fb13c7cb3 into 0ff9c0132a8914d6756739a7a3b085e47870b93d
```

`docs-preview-deploy` then [triggers for commit `0ff9c01`](https://github.com/docker-mailserver/docker-mailserver/actions/runs/11760837244), and the `if` condition above evaluates to false, skipping the workflow 🤔 

---

The `doc-preview-*` workflows were triggered by a [merge commit from `master` into the PR](https://github.com/docker-mailserver/docker-mailserver/pull/4183#issuecomment-2466529988) (_that I did via Github's Web UI button to update the branch_).

[Commit `a5682e7`](https://github.com/docker-mailserver/docker-mailserver/pull/4183/commits/a5682e7c805397493aec6e5f333b236fb13c7cb3) reflects that, while `a721146` would be the pre-merge commit Github created for the PR sync event, and [commit `0ff9c01`](https://github.com/docker-mailserver/docker-mailserver/commit/0ff9c0132a8914d6756739a7a3b085e47870b93d) is the current latest commit on `master` branch (_the squash merge commit from this very merged PR updating the workflow_).

Looking at an [older workflow run](https://github.com/docker-mailserver/docker-mailserver/actions/runs/11091863613) that did pass, [this commit](https://github.com/docker-mailserver/docker-mailserver/commit/3937e1e719f4058e0a4a846bff39383071f8391d) was the trigger, even though it shouldn't have met the [`docs-preview-prepare` path requirement to trigger](https://github.com/docker-mailserver/docker-mailserver/blob/3937e1e719f4058e0a4a846bff39383071f8391d/.github/workflows/docs-preview-prepare.yml#L3-L8), it was the [PR merge to master branch commit](https://github.com/docker-mailserver/docker-mailserver/pull/4201#event-14444896433), rather than a pre-merge commit, or as mentioned in this case a commit merging `master` into the PR branch (_which has been a [valid trigger in the past too](https://github.com/docker-mailserver/docker-mailserver/pull/4183#issuecomment-2361414208), unrelated to the commit content_).

---

The previous working `if` condition was:

```yaml
if: ${{ github.event.workflow_run.event == 'pull_request' && github.event.workflow_run.conclusion == 'success' }}
```

Presumably the `contains` addition is not compatible here. If that's the case then the context may not be valid below either 🤔 

I suppose it depends on if `a721146` (github generated pre-merge commit) or the actual latest commit on the PR branch (`a5682e7`?) is treated as the `head_sha`. Perhaps in this case it's a compatibility issue with the `master` to PR merge commit only, I'll verify that shortly 😕

Reply:
**UPDATE:**
- The condition did not skip on my own test repo when I created a PR and sync'd to changes of the primary branch `main`.
- [Opened a PR](https://github.com/docker-mailserver/docker-mailserver/pull/4258) with a small change to trigger the workflow, that [worked fine](https://github.com/docker-mailserver/docker-mailserver/actions/runs/11761492789).

Perhaps it has something to do with the the failing PR having been opened prior to this CI workflow update, or it's something related to a third-party contributor/branch 🤷‍♂️ 

---

**UPDATE 2:**

[Opened another PR](https://github.com/docker-mailserver/docker-mailserver/pull/4259), this time from my own fork of DMS. This triggered the same problem with [the `if` condition skipping the workflow](https://github.com/docker-mailserver/docker-mailserver/actions/runs/11761829712), and that was just the PR commits itself, no new pushes/updates.

Reply:
I removed the third condition:

```yaml
&& contains(github.event.workflow_run.pull_requests.*.head.sha, github.event.workflow_run.head_sha)
```

The workflow runs but [shows that the context we want is missing completely](https://github.com/docker-mailserver/docker-mailserver/actions/runs/11762692806/job/32765833419#step:4:1):

```
  PR_NUMBER=$(jq -r '[.[] | select(.head.sha == "45f63b686fb61d85fec2ddab7ac4504f8b191555")][0].number' <<< "${pull_requests}")
  {
    echo "PR_NUMBER=${PR_NUMBER}"
    echo 'PR_HEADSHA=45f63b686fb61d85fec2ddab7ac4504f8b191555'
  } >> "${GITHUB_ENV}"
  shell: /usr/bin/bash -e {0}
  env:
    head_sha: 45f63b686fb61d85fec2ddab7ac4504f8b191555
    pull_requests: []
```

Whereas for the working PR (branch on the same repo) [we have that extra context](https://github.com/docker-mailserver/docker-mailserver/actions/runs/11761492789/job/32763286908#step:4:2) available:

```
  PR_NUMBER=$(jq -r '[.[] | select(.head.sha == "8bef84e53657709eda5d35729399817035362efd")][0].number' <<< "${pull_requests}")
  {
    echo "PR_NUMBER=${PR_NUMBER}"
    echo 'PR_HEADSHA=8bef84e53657709eda5d35729399817035362efd'
  } >> "${GITHUB_ENV}"
  shell: /usr/bin/bash -e {0}
  env:
    head_sha: 8bef84e53657709eda5d35729399817035362efd
    pull_requests: [
    {
      "base": {
        "ref": "master",
        "repo": {
          "id": 33037215,
          "name": "docker-mailserver",
          "url": "https://api.github.com/repos/docker-mailserver/docker-mailserver"
        },
        "sha": "0ff9c0132a8914d6756739a7a3b085e47870b93d"
      },
      "head": {
        "ref": "docs/smtp-bind-fix-snippet-titles",
        "repo": {
          "id": 33037215,
          "name": "docker-mailserver",
          "url": "https://api.github.com/repos/docker-mailserver/docker-mailserver"
        },
        "sha": "8bef84e53657709eda5d35729399817035362efd"
      },
      "id": 2171261491,
      "number": 4258,
      "url": "https://api.github.com/repos/docker-mailserver/docker-mailserver/pulls/4258"
    }
  ]
```

So perhaps some metadata does need to be passed through, but then anyone could adjust the PR/issue provided as input 🤷‍♂️

Reply:
Sorry for the late response, I totally missed the notification.
The third condition can be safely removed.
I dont know why `github.event.workflow_run.pull_requests` is empty for that case, but I dont see how passing it via an env var rather than interpolating it directly could change the value it receives from the runner.
I found this answer that says that `pull_requests` is not filled for PRs coming from a fork and they had to iterate through all open PRs to find the right one: https://stackoverflow.com/a/79017997

Reply:
Here we go, from the [GHA workflow event trigger docs](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request)

> _The `pull_request` webhook event payload is empty for merged pull requests and pull requests that come from forked repositories._

---

There has also been an ongoing [discussion about this particular issue for years](https://github.com/orgs/community/discussions/25220#discussioncomment-3246921):
1. [GH blog advises with passing PR number via artifact as ENV from untrusted to trusted workflows](https://github.com/orgs/community/discussions/25220#discussioncomment-7856118) (_which led to the security concern in the first place_ 😅 )
2. [GH CLI or REST API via JS script](https://github.com/orgs/community/discussions/25220#discussioncomment-8697399) (rate limit risk)
3. [`pull_request_target` + re-usable workflows with restricted permissions for untrusted builds](https://github.com/orgs/community/discussions/25220#discussioncomment-9164547)
4. [`gh run view`](https://github.com/orgs/community/discussions/25220#discussioncomment-7532132) provides the PR number in an awkward manner, but is only compatible with forks apparently 🤷‍♂️ 

Perhaps this is something GH can properly address, as community discussions like that with the confusion and various solutions probably isn't benefiting anyone 😅 

---

The [blog post referenced](https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/) by solution 1 above, does touch on `pull_request_target` caveats:

> You may ask yourself: if the `pull_request_target` workflow only checks out and builds the PR, i.e. runs untrusted code but doesn’t reference any secrets, is it still vulnerable?
> 
> Yes it is, because a workflow triggered on `pull_request_target` still has the read/write repository token in memory that is potentially available to any running program.
> If the workflow uses `actions/checkout` and does not pass the optional parameter `persist-credentials` as false, it makes it even worse. The default for the parameter is `true`.
> 
>  It means that in any subsequent steps any running code can simply read the stored repository token from the disk.
>  If you don’t need a repository write access or secrets, just stick to the `pull_request` trigger.

The `persist-credentials` setting appears to have that related logic [here](https://github.com/actions/checkout/blob/3b9b8c884f6b4bb4d5be2779c26374abadae0871/src/git-auth-helper.ts#L151-L189).

The blog post only refers to `steps` for a job, while [solution 3](https://github.com/orgs/community/discussions/25220#discussioncomment-9170862) from above shows a workflow with two jobs, the `build` job of the 1st workflow provides inputs into the 2nd via workflow call trigger to perform a git checkout of the PR without secrets or permissions, which should avoid the `persist-credentials` concern for their `preview` job that runs next?

---

**EDIT:** I've been looking into solution 3 (`pull_request_target`) and it seems like it might be the right way to go, I'll get a PR up for this and ping you for a glance over? :)

Reply:
> The third condition can be safely removed.

Yeah, it was more of a safe-guard and in this case it did correctly skip the workflow because the expected metadata was missing. I dropped it to verify that it was in fact lacking the metadata needed, so a different solution is required.

> I dont know why `github.event.workflow_run.pull_requests` is empty for that case, but I dont see how passing it via an env var rather than interpolating it directly could change the value it receives from the runner.

The previous solution that did work was using the untrusted `pull_request` workflow (prepare) to get the pull request number and related info, then send that over to the trusted `workflow_run` workflow (deploy) which appended the lines to the filepath of `$GITHUB_ENV`.

While that did work, there was a security concern with `LD_PRELOAD` you mentioned and there was various concerns I had with how to approach that properly that I tried the alternative approach that we have in place currently.

I'd rather not go back to the previous method since the PR contributor can manipulate what those values would be.

> I found this answer that says that `pull_requests` is not filled for PRs coming from a fork and they had to iterate through all open PRs to find the right one: https://stackoverflow.com/a/79017997

Yes, as my prior message posted at roughly the same time as your comment notes, that appears to be the case. I'll push ahead with the `pull_request_target` solution 👍